src/util.rs

```diff
@@ -6,22 +6,62 @@ use std::str;
 
 use memchr::{memchr, memchr2};
 
-#[cfg(feature = "compression")]
-use zip::result::ZipError;
-
 #[derive(Clone, Debug, PartialEq)]
-pub enum ParseError {
+pub enum ParseErrorType {
+    BadCompression,
     PrematureEOF,
-    Invalid(String),
+    InvalidHeader,
+    InvalidRecord,
+    IOError,
+    Invalid,
+}
+
+#[derive(Clone, Debug, PartialEq)]
+pub struct ParseError {
+    pub record: usize,
+    pub context: String,
+    pub msg: String,
+    pub error_type: ParseErrorType,
+}
+
+impl ParseError {
+    pub fn new<S>(msg: S, error_type: ParseErrorType) -> Self
+    where
+        S: Into<String>,
+    {
+        ParseError {
+            record: 0,
+            context: "".to_string(),
+            msg: msg.into(),
+            error_type,
+        }
+    }
+
+    pub fn record(mut self, record_number: usize) -> Self {
+        self.record = record_number;
+        self
+    }
+
+    pub fn context<S>(mut self, context: S) -> Self
```

Comment:
I believe you can write `context: impl ToString` as long as `context` impl Display